### ISSUE TITLE
programs using opencl fail with "Could not create topdir for cache"

### ISSUE THREAD

**Issue by xpue** [pocl/pocl#289]
Happens for me on pocl 0.12 with luxmark, web search shows also this: https://developer.blender.org/T46799

**Comment by pjaaskel** [comment]
In your link it says it's fixed in Blender? https://developer.blender.org/rBbfaebeaee64399e4c6da65b8aaa63e1a3cc3d000

**Comment by illwieckz** [comment]
I'm facing the same bug while trying to run `luxmark` on pocl. By the way, since `clinfo` is able to create the cache dir and fill it but `luxmark` is only able to create the cache dir without filling it reveals the bug only appear with some programs (`luxmark`, `blender`…) and not all.

The fact blender users [workaround that bug](https://developer.blender.org/T46799#347478) does not mean it's fixed. People in that blender thread are complaining because installing POCL on their system make blender instable and crashing, even if they will not use POCL. That blender crash happens when people open the settings to chose the OpenCL device to be used by the Cycles rendering engine. Blender crashes because blender test the OpenCL devices before allowing the user to pick one of them, and since there is a bug in pocl, Blender crashes before the user is able to disable the POCL device in Blender.

The workaround found by blender guys is to tell Blender to not test the OpenCL devices automatically so they are able to open the system settings without crashing Blender. In fact on my own system, with latest Blender, disabling that test make all OpenCL devices unavailable, so, due to this bug, I'm only able to run Blender's Cycles engine on CPU since the workaround disables the OpenCL test making the GPU OpenCL devices unavailable. Installing pocl on my system make Blender's Cycles unable to use the GPU.

People in this Blender threads [is saying](https://developer.blender.org/T46799#347476) the bug is in [pocl_cache.c](https://github.com/pocl/pocl/blob/master/lib/CL/pocl_cache.c#L495). I don't know if it's exactly at this line (since the line number is not associated with a commit number, and that link is one year old), but my own experiment made be presume the bug is somewhere near these lines too.

What the [Blender commit](https://developer.blender.org/rBbfaebeaee64399e4c6da65b8aaa63e1a3cc3d000) (referenced in this Cycles thread) do is to exclude some LLVM symbols, this commit explicitly references luxrender so it looks like the bug is known to affect both Cycles and LuxRender. By the way, on my own side, Blender crash. I use Blender 2.78a released 2016-09-30 while this commit was committed 2015-11-27.

If I delete the pocl cache and run clinfo, the cache is created and filled up:

```
$ rm -R ~/.cache/pocl
$ clinfo
$ ls ~/.cache/pocl/kcache/
BG
```

If I run luxmark, it complains about not being able to create the cache dir (and quit) even if the cache already exists:

```
$ rm -R ~/.cache/pocl
$ strace ./luxmark.bin
mkdir("/home/illwieckz/.cache/pocl/kcache", 0770) = -1 EEXIST (File exists)
write(2, "Could not create topdir for cach"..., 33Could not create topdir for cache) = 33
```

and even if it does not exist:

```
$ rm -R ~/.cache/pocl
$ strace ./luxmark.bin
mkdir("/home/illwieckz/.cache/pocl/kcache", 0770) = -1 ENOENT (No such file or directory)
mkdir("/home/illwieckz/.cache/pocl", 0770) = 0
mkdir("/home/illwieckz/.cache/pocl/kcache", 0770) = 0
write(2, "Could not create topdir for cach"..., 33Could not create topdir for cache) = 33
```

same if I use another cache disk:

```
$ export POCL_CACHE_DIR=/tmp/pocl/kcache/
$ strace ./luxmark.bin
access("/tmp/pocl/kcache/", F_OK)       = -1 ENOENT (No such file or directory)
mkdir("/home/illwieckz/.cache/pocl/kcache", 0770) = -1 EEXIST (File exists)
write(2, "Could not create topdir for cach"..., 33Could not create topdir for cache) = 33
```

Since it it looks like user-specified cache dir must be created beforhand, I created it by hand, same result:

```
$ mkdir -p /tmp/pocl/kcache/
access("/tmp/pocl/kcache/", F_OK)       = 0
mkdir("/tmp/pocl/kcache/", 0770)        = -1 EEXIST (File exists)
write(2, "Could not create topdir for cach"..., 33Could not create topdir for cache) = 33
```

I've read #24 and [doc/luxmark.txt](https://github.com/pocl/pocl/blob/master/doc/luxmark.txt) but it's unrelated, first it was about LuxMark 2 and I'm facing this problem with LuxMark 3, second all these `CL_DEVICE_TYPE_CPU=basic` or `POCL_DEVICES=basic` workaround does not work.

That stuff Blender doing with LLVM symbols is interesting, perhaps it's not a truly a pocl fault and perhaps LuxMark is facing the same problem, but… no sorry, it's not fixed.

I tried `pocl` `0.13-7` and `0.13-8` from Ubuntu repository that look to use the exact same source tree. I've seen they apply an llvm-related patch to built them, named `llvm-system-libs.patch`, which this one:

```
Author: Andreas Beckmann <anbe@debian.org>
Description: llvm-config --system-libs is only needed with static llvm

--- a/configure.ac
+++ b/configure.ac
@@ -86,6 +86,7 @@ LLVM_VERSION=`$LLVM_CONFIG --version`
 LLVM_BINDIR=`$LLVM_CONFIG --bindir`
 LLVM_LIBDIR=`$LLVM_CONFIG --libdir`
 LLVM_LDFLAGS=`$LLVM_CONFIG --ldflags`
+LLVM_SYSTEM_LIBS=`$LLVM_CONFIG --system-libs`
 
 AC_SUBST([LLVM_VERSION], [$LLVM_VERSION])
 AC_DEFINE_UNQUOTED([LLVM_VERSION], ["$LLVM_VERSION"], "LLVM version as a string.")
@@ -93,11 +94,9 @@ AC_DEFINE_UNQUOTED([LLVM_VERSION], ["$LL
 case "$LLVM_VERSION" in
      3.7*)
      AC_DEFINE([LLVM_3_7], [], "Using LLVM 3.7")
-     LLVM_LDFLAGS="$LLVM_LDFLAGS `$LLVM_CONFIG --system-libs`"
    ;;
      3.8*)
      AC_DEFINE([LLVM_3_8], [], "Using LLVM 3.8")
-     LLVM_LDFLAGS="$LLVM_LDFLAGS `$LLVM_CONFIG --system-libs`"
    ;;
      *)
    AC_MSG_ERROR(
@@ -122,6 +121,7 @@ AC_ARG_ENABLE([static-llvm],
 LLVM_SHARED_LIB_FILE=$LLVM_LIBDIR/libLLVM-$LLVM_VERSION$LIBRARY_SUFFIX
 
 if test "$link_llvm_static" = "yes"; then
+  LLVM_LDFLAGS="$LLVM_LDFLAGS $LLVM_SYSTEM_LIBS"
   LLVM_LIBS=$($LLVM_CONFIG --libfiles)
 else
   LLVM_LIBS=-lLLVM-$LLVM_VERSION
```

So, I really don't know where is the bug, in Blender? in LuxMark? in LLVM? in POCL? but for sure it's not fixed.

Tested with Blender `2.78a` (Ubuntu build), LuxMark `3.0` (LuxMark build), POCL `0.13` (Ubuntu build).

**Comment by pjaaskel** [comment]
Please try again with pocl master.

**Comment by franz** [comment]
> People in this Blender threads is saying the bug is in pocl_cache.c. 
> my own experiment made be presume the bug is somewhere near these lines too.

"People" are not reading pocl code very carefully... that line
https://github.com/pocl/pocl/blob/master/lib/CL/pocl_cache.c#L565

... is using this code:
https://github.com/pocl/pocl/blob/master/lib/llvmopencl/LLVMFileUtils.cc#L122

.. which in turn is calling `sys::fs::create_directories(p, true)`. The second argument (true) is for ignore-existing.

What i have found out is, if i run the luxmark binary downloaded from the Luxmark webpage, it crashes; but if i run Luxmark from distro package, it launches the GUI (it crashes on something else later, but that's an unrelated bug). The problem is that Luxmark is shipping binaries which they only mark as "Linux 64bit", so people assume they work on every 64bit linux distro. Clearly they don't.

> What the Blender commit (referenced in this Cycles thread) do is to exclude some LLVM symbols

That *looks* like it might be the fix (which needs to be applied to Luxmark). But it might also be just a minor step in getting it work. My own experience with linux programs distributed as binaries tells me the only way to get it working reliably is to ship everything starting from glibc (Adobe & Valve do that too, and here you see why).

In short: always install Luxmark from your distro's packages.
